routemaster/state_machine/api.py

```diff
@@ -257,9 +247,6 @@ def process_cron(
         relevant_labels = get_labels(state_machine, state, conn)
 
     for label_name in relevant_labels:
```

Comment:
Given the way that we know the label lookups work -- that they're a query which is then wrapped in an `itertools.takewhile` that uses the `is_terminating` callback to filter the results (in `_iter_labels_until_terminating`), I'm not sure that it's a good idea to use the results outside the transaction which queried them.

While I _think_ our current usage is safe, it's entirely non-obvious that that's the case (since each layer returns an `Iterable[str]` that _happens_ to be a `List[str]`) and relies on a lot of knowledge of the internals.

I definitely see the benefit of @prophile's suggesting that the `is_terminating` handling be in that callable, though I'm not sure it's worth the trade-off of the complexity around the correctness of the queries that it incurs.

Reply:
I've changed this to enforce returning a `List[str]` instead of an iterable.